A private role with no users breaks /api/roles + /api/roles times out

On release_24.2 following our update, https://usegalaxy.org.au/api/roles leads to 'Internal Server Error'.

Stack trace ends in `Did not find exactly one user for private role <galaxy.model.Role(1) at 0x7f36a0e1fa30>` - I'll add this as a comment because it's long.

At first we were looking for a role with more than one user_role_association, then Nuwan worked out that it could be a role with no user_role_assocation - we found this and updated its type.

We think this might have solved the problem but it's not clear because the api now times out when loading. Setting `?limit=` or `?page=` makes no difference

```
Feb 19 03:40:55 galaxy galaxyctl[2622924]: [2025-02-19 03:40:55 +0000] [2622924] [ERROR] Exception in ASGI application
Feb 19 03:40:55 galaxy galaxyctl[2622924]: Traceback (most recent call last):
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/venv/lib/python3.10/site-packages/uvicorn/protocols/http/h11_impl.py", line 406, in run_asgi
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     result = await app(  # type: ignore[func-returns-value]
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/venv/lib/python3.10/site-packages/uvicorn/middleware/proxy_headers.py", line 60, in __call__
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     return await self.app(scope, receive, send)
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/venv/lib/python3.10/site-packages/fastapi/applications.py", line 1054, in __call__
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     await super().__call__(scope, receive, send)
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/venv/lib/python3.10/site-packages/sentry_sdk/integrations/starlette.py", line 409, in _sentry_patched_asgi_app
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     return await middleware(scope, receive, send)
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/venv/lib/python3.10/site-packages/sentry_sdk/integrations/asgi.py", line 161, in _run_asgi3
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     return await self._run_app(scope, receive, send, asgi_version=3)
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/venv/lib/python3.10/site-packages/sentry_sdk/integrations/asgi.py", line 262, in _run_app
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     raise exc from None
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/venv/lib/python3.10/site-packages/sentry_sdk/integrations/asgi.py", line 257, in _run_app
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     return await self.app(
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/venv/lib/python3.10/site-packages/starlette/applications.py", line 113, in __call__
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     await self.middleware_stack(scope, receive, send)
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/venv/lib/python3.10/site-packages/sentry_sdk/integrations/starlette.py", line 200, in _create_span_call
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     return await old_call(app, scope, new_receive, new_send, **kwargs)
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/venv/lib/python3.10/site-packages/starlette/middleware/errors.py", line 187, in __call__
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     raise exc
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/venv/lib/python3.10/site-packages/starlette/middleware/errors.py", line 165, in __call__
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     await self.app(scope, receive, _send)
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/venv/lib/python3.10/site-packages/sentry_sdk/integrations/starlette.py", line 200, in _create_span_call
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     return await old_call(app, scope, new_receive, new_send, **kwargs)
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/venv/lib/python3.10/site-packages/starlette_context/middleware/raw_middleware.py", line 92, in __call__
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     await self.app(scope, receive, send_wrapper)
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/venv/lib/python3.10/site-packages/sentry_sdk/integrations/starlette.py", line 200, in _create_span_call
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     return await old_call(app, scope, new_receive, new_send, **kwargs)
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/venv/lib/python3.10/site-packages/starlette/middleware/base.py", line 185, in __call__
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     with collapse_excgroups():
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/usr/lib/python3.10/contextlib.py", line 153, in __exit__
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     self.gen.throw(typ, value, traceback)
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/venv/lib/python3.10/site-packages/starlette/_utils.py", line 82, in collapse_excgroups
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     raise exc
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/venv/lib/python3.10/site-packages/starlette/middleware/base.py", line 187, in __call__
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     response = await self.dispatch_func(request, call_next)
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/galaxy-app/lib/galaxy/webapps/galaxy/fast_app.py", line 108, in add_x_frame_options
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     response = await call_next(request)
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/venv/lib/python3.10/site-packages/starlette/middleware/base.py", line 163, in call_next
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     raise app_exc
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/venv/lib/python3.10/site-packages/starlette/middleware/base.py", line 149, in coro
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     await self.app(scope, receive_or_disconnect, send_no_error)
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/venv/lib/python3.10/site-packages/sentry_sdk/integrations/starlette.py", line 298, in _sentry_exceptionmiddleware_call
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     await old_call(self, scope, receive, send)
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/venv/lib/python3.10/site-packages/sentry_sdk/integrations/starlette.py", line 200, in _create_span_call
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     return await old_call(app, scope, new_receive, new_send, **kwargs)
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/venv/lib/python3.10/site-packages/starlette/middleware/exceptions.py", line 62, in __call__
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     await wrap_app_handling_exceptions(self.app, conn)(scope, receive, send)
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/venv/lib/python3.10/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     raise exc
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/venv/lib/python3.10/site-packages/starlette/_exception_handler.py", line 42, in wrapped_app
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     await app(scope, receive, sender)
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/venv/lib/python3.10/site-packages/starlette/routing.py", line 715, in __call__
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     await self.middleware_stack(scope, receive, send)
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/venv/lib/python3.10/site-packages/starlette/routing.py", line 735, in app
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     await route.handle(scope, receive, send)
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/venv/lib/python3.10/site-packages/starlette/routing.py", line 288, in handle
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     await self.app(scope, receive, send)
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/venv/lib/python3.10/site-packages/starlette/routing.py", line 76, in app
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     await wrap_app_handling_exceptions(app, request)(scope, receive, send)
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/venv/lib/python3.10/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     raise exc
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/venv/lib/python3.10/site-packages/starlette/_exception_handler.py", line 42, in wrapped_app
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     await app(scope, receive, sender)
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/venv/lib/python3.10/site-packages/starlette/routing.py", line 73, in app
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     response = await f(request)
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/venv/lib/python3.10/site-packages/sentry_sdk/integrations/fastapi.py", line 143, in _sentry_app
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     return await old_app(*args, **kwargs)
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/venv/lib/python3.10/site-packages/fastapi/routing.py", line 301, in app
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     raw_response = await run_endpoint_function(
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/venv/lib/python3.10/site-packages/fastapi/routing.py", line 214, in run_endpoint_function
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     return await run_in_threadpool(dependant.call, **values)
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/venv/lib/python3.10/site-packages/starlette/concurrency.py", line 39, in run_in_threadpool
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     return await anyio.to_thread.run_sync(func, *args)
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/venv/lib/python3.10/site-packages/anyio/to_thread.py", line 56, in run_sync
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     return await get_async_backend().run_sync_in_worker_thread(
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/venv/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 2364, in run_sync_in_worker_thread
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     return await future
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/venv/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 864, in run
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     result = context.run(func, *args)
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/venv/lib/python3.10/site-packages/sentry_sdk/integrations/fastapi.py", line 99, in _sentry_call
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     return old_call(*args, **kwargs)
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/galaxy-app/lib/galaxy/webapps/galaxy/api/roles.py", line 37, in index
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     return self.service.get_index(trans=trans)
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/galaxy-app/lib/galaxy/webapps/galaxy/services/roles.py", line 36, in get_index
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     return RoleListResponse(root=[role_to_model(r) for r in roles])
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/galaxy-app/lib/galaxy/webapps/galaxy/services/roles.py", line 36, in <listcomp>
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     return RoleListResponse(root=[role_to_model(r) for r in roles])
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/galaxy-app/lib/galaxy/webapps/galaxy/services/roles.py", line 18, in role_to_model
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     item = role.to_dict(view="element")
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/galaxy-app/lib/galaxy/util/dictifiable.py", line 92, in to_dict
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     return self._dictify_view_keys(view=view, value_mapper=value_mapper)
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/galaxy-app/lib/galaxy/util/dictifiable.py", line 70, in _dictify_view_keys
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     item = self.__getattribute__(key)
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/venv/lib/python3.10/site-packages/sqlalchemy/ext/hybrid.py", line 1120, in __get__
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     return self.fget(instance)
Feb 19 03:40:55 galaxy galaxyctl[2622924]:   File "/mnt/galaxy/galaxy-app/lib/galaxy/model/__init__.py", line 3850, in name
Feb 19 03:40:55 galaxy galaxyctl[2622924]:     assert len(user_assocs) == 1, f"Did not find exactly one user for private role {self}"
Feb 19 03:40:55 galaxy galaxyctl[2622924]: AssertionError: Did not find exactly one user for private role <galaxy.model.Role(1) at 0x7f36a0e1fa30>
```

--- a/lib/galaxy/files/__init__.py
+++ b/lib/galaxy/files/__init__.py
@@ -396,7 +396,11 @@ class ProvidesFileSourcesUserContext(FileSourcesUserContext, FileSourceDictifiab
     def role_names(self) -> Set[str]:
         """The set of role names of this user."""
         user = self.trans.user
-        return {ura.role.name for ura in user.roles} if user else set()
+        role_names = set()
+        if user:
+            role_names = {ura.role.name for ura in user.roles}
+            role_names.add(user.email)  # User's private role may have a generic name, so add user's email explicitly.
+        return role_names
 
     @property
     def group_names(self) -> Set[str]:

--- a/lib/galaxy/managers/datasets.py
+++ b/lib/galaxy/managers/datasets.py
@@ -11,6 +11,7 @@ from typing import (
     Generic,
     List,
     Optional,
+    Set,
     Type,
     TypeVar,
 )
@@ -36,6 +37,7 @@ from galaxy.model import (
     HistoryDatasetAssociation,
 )
 from galaxy.model.base import transaction
+from galaxy.model.db.role import get_private_role_user_emails_dict
 from galaxy.schema.tasks import (
     ComputeDatasetHashTaskRequest,
     PurgeDatasetsTaskRequest,
@@ -451,16 +453,24 @@ class DatasetAssociationManager(
             library_dataset = None
             dataset = dataset_assoc.dataset
 
+        private_role_emails = get_private_role_user_emails_dict(self.session())
+
         # Omit duplicated roles by converting to set
         access_roles = set(dataset.get_access_roles(self.app.security_agent))
         manage_roles = set(dataset.get_manage_permissions_roles(self.app.security_agent))
 
-        access_dataset_role_list = [
-            (access_role.name, self.app.security.encode_id(access_role.id)) for access_role in access_roles
-        ]
-        manage_dataset_role_list = [
-            (manage_role.name, self.app.security.encode_id(manage_role.id)) for manage_role in manage_roles
-        ]
+        def make_tuples(roles: Set):
+            tuples = []
+            for role in roles:
+                # use role name for non-private roles, and user.email from private rules
+                displayed_name = private_role_emails.get(role.id, role.name)
+                role_tuple = (displayed_name, self.app.security.encode_id(role.id))
+                tuples.append(role_tuple)
+            return tuples
+
+        access_dataset_role_list = make_tuples(access_roles)
+        manage_dataset_role_list = make_tuples(manage_roles)
+
         rval = dict(access_dataset_roles=access_dataset_role_list, manage_dataset_roles=manage_dataset_role_list)
         if library_dataset is not None:
             modify_roles = set(
@@ -468,9 +478,7 @@ class DatasetAssociationManager(
                     library_dataset, self.app.security_agent.permitted_actions.LIBRARY_MODIFY
                 )
             )
-            modify_item_role_list = [
-                (modify_role.name, self.app.security.encode_id(modify_role.id)) for modify_role in modify_roles
-            ]
+            modify_item_role_list = make_tuples(modify_roles)
             rval["modify_item_roles"] = modify_item_role_list
         return rval
 

--- a/lib/galaxy/managers/folders.py
+++ b/lib/galaxy/managers/folders.py
@@ -7,6 +7,7 @@ from dataclasses import dataclass
 from typing import (
     List,
     Optional,
+    Set,
     Tuple,
     Union,
 )
@@ -49,6 +50,7 @@ from galaxy.model import (
     LibraryFolderPermissions,
 )
 from galaxy.model.base import transaction
+from galaxy.model.db.role import get_private_role_user_emails_dict
 from galaxy.model.scoped_session import galaxy_scoped_session
 from galaxy.schema.schema import LibraryFolderContentsIndexQueryPayload
 from galaxy.security import RBACAgent
@@ -295,6 +297,8 @@ class FolderManager:
         :returns:   dict of current roles for all available permission types
         :rtype:     dictionary
         """
+        private_role_emails = get_private_role_user_emails_dict(trans.sa_session)
+
         # Omit duplicated roles by converting to set
         modify_roles = set(
             trans.app.security_agent.get_roles_for_action(
@@ -312,17 +316,19 @@ class FolderManager:
             )
         )
 
-        modify_folder_role_list = [
-            (modify_role.name, trans.security.encode_id(modify_role.id)) for modify_role in modify_roles
-        ]
-        manage_folder_role_list = [
-            (manage_role.name, trans.security.encode_id(manage_role.id)) for manage_role in manage_roles
-        ]
-        add_library_item_role_list = [(add_role.name, trans.security.encode_id(add_role.id)) for add_role in add_roles]
+        def make_tuples(roles: Set):
+            tuples = []
+            for role in roles:
+                # use role name for non-private roles, and user.email from private rules
+                displayed_name = private_role_emails.get(role.id, role.name)
+                role_tuple = (displayed_name, trans.security.encode_id(role.id))
+                tuples.append(role_tuple)
+            return tuples
+
         return dict(
-            modify_folder_role_list=modify_folder_role_list,
-            manage_folder_role_list=manage_folder_role_list,
-            add_library_item_role_list=add_library_item_role_list,
+            modify_folder_role_list=make_tuples(modify_roles),
+            manage_folder_role_list=make_tuples(manage_roles),
+            add_library_item_role_list=make_tuples(add_roles),
         )
 
     def can_add_item(self, trans, folder):

--- a/lib/galaxy/managers/libraries.py
+++ b/lib/galaxy/managers/libraries.py
@@ -30,6 +30,7 @@ from galaxy.model.db.library import (
     get_library_ids,
     get_library_permissions_by_role,
 )
+from galaxy.model.db.role import get_private_role_user_emails_dict
 from galaxy.util import (
     pretty_print_time_interval,
     unicodify,
@@ -277,26 +278,26 @@ class LibraryManager:
         :rtype:     dictionary
         :returns:   dict of current roles for all available permission types
         """
-        access_library_role_list = [
-            (access_role.name, trans.security.encode_id(access_role.id))
-            for access_role in self.get_access_roles(trans, library)
-        ]
-        modify_library_role_list = [
-            (modify_role.name, trans.security.encode_id(modify_role.id))
-            for modify_role in self.get_modify_roles(trans, library)
-        ]
-        manage_library_role_list = [
-            (manage_role.name, trans.security.encode_id(manage_role.id))
-            for manage_role in self.get_manage_roles(trans, library)
-        ]
-        add_library_item_role_list = [
-            (add_role.name, trans.security.encode_id(add_role.id)) for add_role in self.get_add_roles(trans, library)
-        ]
+        private_role_emails = get_private_role_user_emails_dict(trans.sa_session)
+        access_roles = self.get_access_roles(trans, library)
+        modify_roles = self.get_modify_roles(trans, library)
+        manage_roles = self.get_manage_roles(trans, library)
+        add_roles = self.get_add_roles(trans, library)
+
+        def make_tuples(roles: Set):
+            tuples = []
+            for role in roles:
+                # use role name for non-private roles, and user.email from private rules
+                displayed_name = private_role_emails.get(role.id, role.name)
+                role_tuple = (displayed_name, trans.security.encode_id(role.id))
+                tuples.append(role_tuple)
+            return tuples
+
         return dict(
-            access_library_role_list=access_library_role_list,
-            modify_library_role_list=modify_library_role_list,
-            manage_library_role_list=manage_library_role_list,
-            add_library_item_role_list=add_library_item_role_list,
+            access_library_role_list=make_tuples(access_roles),
+            modify_library_role_list=make_tuples(modify_roles),
+            manage_library_role_list=make_tuples(manage_roles),
+            add_library_item_role_list=make_tuples(add_roles),
         )
 
     def get_access_roles(self, trans, library: Library) -> Set[Role]:

--- a/lib/galaxy/managers/roles.py
+++ b/lib/galaxy/managers/roles.py
@@ -84,11 +84,7 @@ class RoleManager(base.ModelManager[model.Role]):
         user_ids = role_definition_model.user_ids or []
         group_ids = role_definition_model.group_ids or []
 
-        stmt = (
-            select(Role)
-            .where(Role.name == name)  # type:ignore[arg-type,comparison-overlap]  # Role.name is a SA hybrid property
-            .limit(1)
-        )
+        stmt = select(Role).where(Role.name == name).limit(1)
         if trans.sa_session.scalars(stmt).first():
             raise Conflict(f"A role with that name already exists [{name}]")
 

--- a/lib/galaxy/managers/roles.py
+++ b/lib/galaxy/managers/roles.py
@@ -5,10 +5,7 @@ Manager and Serializer for Roles.
 import logging
 from typing import List
 
-from sqlalchemy import (
-    false,
-    select,
-)
+from sqlalchemy import select
 from sqlalchemy.exc import (
     MultipleResultsFound,
     NoResultFound,
@@ -26,6 +23,7 @@ from galaxy.managers import base
 from galaxy.managers.context import ProvidesUserContext
 from galaxy.model import Role
 from galaxy.model.base import transaction
+from galaxy.model.db.role import get_displayable_roles
 from galaxy.schema.schema import RoleDefinitionModel
 from galaxy.util import unicodify
 
@@ -71,12 +69,7 @@ class RoleManager(base.ModelManager[model.Role]):
         return role
 
     def list_displayable_roles(self, trans: ProvidesUserContext) -> List[Role]:
-        roles = []
-        stmt = select(Role).where(Role.deleted == false())
-        for role in trans.sa_session.scalars(stmt):
-            if trans.user_is_admin or trans.app.security_agent.ok_to_display(trans.user, role):
-                roles.append(role)
-        return roles
+        return get_displayable_roles(trans.sa_session, trans.user, trans.user_is_admin, trans.app.security_agent)
 
     def create_role(self, trans: ProvidesUserContext, role_definition_model: RoleDefinitionModel) -> model.Role:
         name = role_definition_model.name

--- a/lib/galaxy/model/__init__.py
+++ b/lib/galaxy/model/__init__.py
@@ -112,7 +112,6 @@ from sqlalchemy.ext.associationproxy import (
     association_proxy,
     AssociationProxy,
 )
-from sqlalchemy.ext.hybrid import hybrid_property
 from sqlalchemy.ext.orderinglist import ordering_list
 from sqlalchemy.orm import (
     aliased,
@@ -3836,7 +3835,7 @@ class Role(Base, Dictifiable, RepresentById):
     id: Mapped[int] = mapped_column(primary_key=True)
     create_time: Mapped[datetime] = mapped_column(default=now, nullable=True)
     update_time: Mapped[datetime] = mapped_column(default=now, onupdate=now, nullable=True)
-    _name: Mapped[str] = mapped_column("name", String(255), index=True)
+    name: Mapped[str] = mapped_column(String(255), index=True)
     description: Mapped[Optional[str]] = mapped_column(TEXT)
     type: Mapped[Optional[str]] = mapped_column(String(40), index=True)
     deleted: Mapped[Optional[bool]] = mapped_column(index=True, default=False)
@@ -3858,19 +3857,6 @@ class Role(Base, Dictifiable, RepresentById):
     @staticmethod
     def default_name(role_type):
         return f"{role_type.value} role"
-
-    @hybrid_property
-    def name(self):
-        if self.type == Role.types.PRIVATE:
-            user_assocs = self.users
-            assert len(user_assocs) == 1, f"Did not find exactly one user for private role {self}"
-            return user_assocs[0].user.email
-        else:
-            return self._name
-
-    @name.setter  # type:ignore[no-redef]  # property setter
-    def name(self, name):
-        self._name = name
 
     def __init__(self, name=None, description=None, type=types.SYSTEM, deleted=False):
         self.name = name or Role.default_name(type)

--- a/lib/galaxy/model/db/role.py
+++ b/lib/galaxy/model/db/role.py
@@ -1,3 +1,5 @@
+from typing import Dict
+
 from sqlalchemy import (
     and_,
     false,
@@ -6,6 +8,7 @@ from sqlalchemy import (
 
 from galaxy.model import (
     Role,
+    User,
     UserRoleAssociation,
 )
 from galaxy.model.scoped_session import galaxy_scoped_session
@@ -50,3 +53,10 @@ def get_displayable_roles(session, trans_user, user_is_admin, security_agent):
         if user_is_admin or security_agent.ok_to_display(trans_user, role):
             roles.append(role)
     return roles
+
+
+def get_private_role_user_emails_dict(session) -> Dict[int, str]:
+    """Return a mapping of private role ids to user emails."""
+    stmt = select(UserRoleAssociation.role_id, User.email).join(Role).join(User).where(Role.type == Role.types.PRIVATE)
+    roleid_email_tuples = session.execute(stmt).all()
+    return dict(roleid_email_tuples)

--- a/lib/galaxy/model/db/role.py
+++ b/lib/galaxy/model/db/role.py
@@ -41,3 +41,12 @@ def get_private_user_role(user, session):
 def get_roles_by_ids(session: galaxy_scoped_session, role_ids):
     stmt = select(Role).where(Role.id.in_(role_ids))
     return session.scalars(stmt).all()
+
+
+def get_displayable_roles(session, trans_user, user_is_admin, security_agent):
+    roles = []
+    stmt = select(Role).where(Role.deleted == false())
+    for role in session.scalars(stmt):
+        if user_is_admin or security_agent.ok_to_display(trans_user, role):
+            roles.append(role)
+    return roles

--- a/lib/galaxy/webapps/galaxy/api/group_roles.py
+++ b/lib/galaxy/webapps/galaxy/api/group_roles.py
@@ -3,9 +3,11 @@ API operations on Group objects.
 """
 
 import logging
+from typing import Optional
 
 from galaxy.managers.context import ProvidesAppContext
 from galaxy.managers.group_roles import GroupRolesManager
+from galaxy.model.db.role import get_private_role_user_emails_dict
 from galaxy.schema.fields import Security
 from galaxy.schema.schema import (
     GroupRoleListResponse,
@@ -26,11 +28,12 @@ log = logging.getLogger(__name__)
 router = Router(tags=["group_roles"])
 
 
-def group_role_to_model(trans, group_id: int, role) -> GroupRoleResponse:
+def group_role_to_model(trans, group_id: int, role, displayed_name: Optional[str] = None) -> GroupRoleResponse:
     encoded_group_id = Security.security.encode_id(group_id)
     encoded_role_id = Security.security.encode_id(role.id)
     url = trans.url_builder("group_role", group_id=encoded_group_id, role_id=encoded_role_id)
-    return GroupRoleResponse(id=role.id, name=role.name, url=url)
+    displayed_name = displayed_name or role.name
+    return GroupRoleResponse(id=role.id, name=displayed_name, url=url)
 
 
 @router.cbv
@@ -49,7 +52,13 @@ class FastAPIGroupRoles:
         trans: ProvidesAppContext = DependsOnTrans,
     ) -> GroupRoleListResponse:
         group_roles = self.manager.index(trans, group_id)
-        return GroupRoleListResponse(root=[group_role_to_model(trans, group_id, gr.role) for gr in group_roles])
+        private_role_emails = get_private_role_user_emails_dict(trans.sa_session)
+        data = []
+        for group in group_roles:
+            role = group.role
+            displayed_name = private_role_emails.get(role.id, role.name)
+            data.append(group_role_to_model(trans, group_id, role, displayed_name))
+        return GroupRoleListResponse(root=data)
 
     @router.get(
         "/api/groups/{group_id}/roles/{role_id}",

--- a/lib/galaxy/webapps/galaxy/api/library_datasets.py
+++ b/lib/galaxy/webapps/galaxy/api/library_datasets.py
@@ -27,6 +27,7 @@ from galaxy.managers import (
     roles,
 )
 from galaxy.model.base import transaction
+from galaxy.model.db.role import get_private_role_user_emails_dict
 from galaxy.structured_app import StructuredApp
 from galaxy.tools.actions import upload_common
 from galaxy.tools.parameters import populate_state
@@ -150,10 +151,12 @@ class LibraryDatasetsController(BaseGalaxyAPIController, UsesVisualizationMixin,
                 page_limit = 10
             query = kwd.get("q", None)
             roles, total_roles = trans.app.security_agent.get_valid_roles(trans, dataset, query, page, page_limit)
+            private_role_emails = get_private_role_user_emails_dict(trans.sa_session)
             return_roles = []
             for role in roles:
                 role_id = trans.security.encode_id(role.id)
-                return_roles.append(dict(id=role_id, name=role.name, type=role.type))
+                displayed_name = private_role_emails.get(role.id, role.name)
+                return_roles.append(dict(id=role_id, name=displayed_name, type=role.type))
             return dict(roles=return_roles, page=page, page_limit=page_limit, total=total_roles)
         else:
             raise exceptions.RequestParameterInvalidException(

--- a/lib/galaxy/webapps/galaxy/api/users.py
+++ b/lib/galaxy/webapps/galaxy/api/users.py
@@ -42,6 +42,7 @@ from galaxy.model import (
     UserQuotaUsage,
 )
 from galaxy.model.base import transaction
+from galaxy.model.db.role import get_private_role_user_emails_dict
 from galaxy.schema import APIKeyModel
 from galaxy.schema.schema import (
     AnonUserModel,
@@ -1091,7 +1092,16 @@ class UserAPIController(BaseGalaxyAPIController, UsesTagsMixin, BaseUIController
         """
         payload = payload or {}
         user = self._get_user(trans, id)
-        roles = user.all_roles()
+
+        def get_role_tuples():
+            private_role_emails = get_private_role_user_emails_dict(trans.sa_session)
+            role_tuples = set()
+            for role in user.all_roles():
+                displayed_name = private_role_emails.get(role.id, role.name)
+                role_tuples.add((displayed_name, role.id))
+            return list(role_tuples)
+
+        role_tuples = get_role_tuples()
         inputs = []
         for index, action in trans.app.model.Dataset.permitted_actions.items():
             inputs.append(
@@ -1102,7 +1112,7 @@ class UserAPIController(BaseGalaxyAPIController, UsesTagsMixin, BaseUIController
                     "name": index,
                     "label": action.action,
                     "help": action.description,
-                    "options": list({(r.name, r.id) for r in roles}),
+                    "options": role_tuples,
                     "value": [a.role.id for a in user.default_permissions if a.action == action.action],
                 }
             )

--- a/lib/galaxy/webapps/galaxy/controllers/admin.py
+++ b/lib/galaxy/webapps/galaxy/controllers/admin.py
@@ -19,6 +19,7 @@ from galaxy.exceptions import (
 )
 from galaxy.managers.quotas import QuotaManager
 from galaxy.model.base import transaction
+from galaxy.model.db.role import get_private_role_user_emails_dict
 from galaxy.model.index_filter_util import (
     raw_text_column_filter,
     text_column_filter,
@@ -879,6 +880,9 @@ class AdminGalaxy(controller.JSAppLauncher):
                 if user in [x.user for x in group.users]:
                     in_users.append(trans.security.encode_id(user.id))
                 all_users.append((user.email, trans.security.encode_id(user.id)))
+
+            private_role_emails = get_private_role_user_emails_dict(trans.sa_session)
+
             for role in (
                 trans.sa_session.query(trans.app.model.Role)
                 .filter(trans.app.model.Role.deleted == false())
@@ -886,7 +890,8 @@ class AdminGalaxy(controller.JSAppLauncher):
             ):
                 if role in [x.role for x in group.roles]:
                     in_roles.append(trans.security.encode_id(role.id))
-                all_roles.append((role.name, trans.security.encode_id(role.id)))
+                displayed_name = private_role_emails.get(role.id, role.name)
+                all_roles.append((displayed_name, trans.security.encode_id(role.id)))
             return {
                 "title": f"Group '{group.name}'",
                 "message": "Group '%s' is currently associated with %d user(s) and %d role(s)."
@@ -922,12 +927,16 @@ class AdminGalaxy(controller.JSAppLauncher):
                 .order_by(trans.app.model.User.table.c.email)
             ):
                 all_users.append((user.email, trans.security.encode_id(user.id)))
+
+            private_role_emails = get_private_role_user_emails_dict(trans.sa_session)
+
             for role in (
                 trans.sa_session.query(trans.app.model.Role)
                 .filter(trans.app.model.Role.deleted == false())
                 .order_by(trans.app.model.Role.name)
             ):
-                all_roles.append((role.name, trans.security.encode_id(role.id)))
+                displayed_name = private_role_emails.get(role.id, role.name)
+                all_roles.append((displayed_name, trans.security.encode_id(role.id)))
             return {
                 "title": "Create Group",
                 "title_id": "create-group",

--- a/lib/galaxy/webapps/galaxy/controllers/dataset.py
+++ b/lib/galaxy/webapps/galaxy/controllers/dataset.py
@@ -166,7 +166,7 @@ class DatasetInterface(BaseUIController, UsesAnnotations, UsesItemRatings, UsesE
             role_tuples = []
             for role in trans.app.security_agent.get_legitimate_roles(trans, data.dataset, "root"):
                 displayed_name = private_role_emails.get(role.id, role.name)
-                role_tuples.add((displayed_name, trans.security.encode_id(role.id)))
+                role_tuples.append((displayed_name, trans.security.encode_id(role.id)))
 
             data_metadata = list(data.metadata.spec.items())
             converters_collection = [(key, value.name) for key, value in data.get_converter_types().items()]

--- a/lib/galaxy/webapps/galaxy/controllers/dataset.py
+++ b/lib/galaxy/webapps/galaxy/controllers/dataset.py
@@ -30,6 +30,7 @@ from galaxy.managers.hdas import (
 )
 from galaxy.managers.histories import HistoryManager
 from galaxy.model.base import transaction
+from galaxy.model.db.role import get_private_role_user_emails_dict
 from galaxy.model.item_attrs import (
     UsesAnnotations,
     UsesItemRatings,
@@ -160,10 +161,13 @@ class DatasetInterface(BaseUIController, UsesAnnotations, UsesItemRatings, UsesE
                 if dtype_value.is_datatype_change_allowed()
             ]
             ldatatypes.sort()
-            all_roles = [
-                (r.name, trans.security.encode_id(r.id))
-                for r in trans.app.security_agent.get_legitimate_roles(trans, data.dataset, "root")
-            ]
+
+            private_role_emails = get_private_role_user_emails_dict(trans.sa_session)
+            role_tuples = []
+            for role in trans.app.security_agent.get_legitimate_roles(trans, data.dataset, "root"):
+                displayed_name = private_role_emails.get(role.id, role.name)
+                role_tuples.add((displayed_name, trans.security.encode_id(role.id)))
+
             data_metadata = list(data.metadata.spec.items())
             converters_collection = [(key, value.name) for key, value in data.get_converter_types().items()]
             can_manage_dataset = trans.app.security_agent.can_manage_dataset(
@@ -273,7 +277,7 @@ class DatasetInterface(BaseUIController, UsesAnnotations, UsesItemRatings, UsesE
                                 "name": index,
                                 "label": action.action,
                                 "help": help_text,
-                                "options": all_roles,
+                                "options": role_tuples,
                                 "value": in_roles.get(action.action),
                                 "readonly": not can_manage_dataset,
                             }

--- a/lib/galaxy/webapps/galaxy/controllers/history.py
+++ b/lib/galaxy/webapps/galaxy/controllers/history.py
@@ -12,6 +12,7 @@ from galaxy.managers import histories
 from galaxy.managers.sharable import SlugBuilder
 from galaxy.model import Role
 from galaxy.model.base import transaction
+from galaxy.model.db.role import get_private_role_user_emails_dict
 from galaxy.model.item_attrs import (
     UsesAnnotations,
     UsesItemRatings,
@@ -134,13 +135,20 @@ class HistoryController(BaseUIController, SharableMixin, UsesAnnotations, UsesIt
         history = self.history_manager.get_owned(self.decode_id(history_id), trans.user, current_history=trans.history)
         if trans.request.method == "GET":
             inputs = []
-            all_roles = trans.user.all_roles()
+            all_roles = set(trans.user.all_roles())
+            private_role_emails = get_private_role_user_emails_dict(trans.sa_session)
             current_actions = history.default_permissions
             for action_key, action in trans.app.model.Dataset.permitted_actions.items():
                 in_roles = set()
                 for a in current_actions:
                     if a.action == action.action:
                         in_roles.add(a.role)
+
+                role_tuples = []
+                for role in all_roles:
+                    displayed_name = private_role_emails.get(role.id, role.name)
+                    role_tuples.append((displayed_name, trans.security.encode_id(role.id)))
+
                 inputs.append(
                     {
                         "type": "select",
@@ -150,7 +158,7 @@ class HistoryController(BaseUIController, SharableMixin, UsesAnnotations, UsesIt
                         "name": action_key,
                         "label": action.action,
                         "help": action.description,
-                        "options": [(role.name, trans.security.encode_id(role.id)) for role in set(all_roles)],
+                        "options": role_tuples,
                         "value": [trans.security.encode_id(role.id) for role in in_roles],
                     }
                 )

--- a/lib/galaxy/webapps/galaxy/services/libraries.py
+++ b/lib/galaxy/webapps/galaxy/services/libraries.py
@@ -16,6 +16,7 @@ from galaxy.managers.folders import FolderManager
 from galaxy.managers.libraries import LibraryManager
 from galaxy.managers.roles import RoleManager
 from galaxy.model import Role
+from galaxy.model.db.role import get_private_role_user_emails_dict
 from galaxy.schema.fields import DecodedDatabaseIdField
 from galaxy.schema.schema import (
     BasicRoleModel,
@@ -170,9 +171,11 @@ class LibrariesService(ServiceBase, ConsumesModelStores):
                 trans, library, query, page, page_limit, is_library_access
             )
 
+            private_role_emails = get_private_role_user_emails_dict(trans.sa_session)
             return_roles = []
             for role in roles:
-                return_roles.append(BasicRoleModel(id=role.id, name=role.name, type=role.type))
+                displayed_name = private_role_emails.get(role.id, role.name)
+                return_roles.append(BasicRoleModel(id=role.id, name=displayed_name, type=role.type))
             return LibraryAvailablePermissions.model_construct(
                 roles=return_roles, page=page, page_limit=page_limit, total=total_roles
             )

--- a/lib/galaxy/webapps/galaxy/services/library_folders.py
+++ b/lib/galaxy/webapps/galaxy/services/library_folders.py
@@ -12,6 +12,7 @@ from galaxy.exceptions import (
 )
 from galaxy.managers.folders import FolderManager
 from galaxy.managers.roles import RoleManager
+from galaxy.model.db.role import get_private_role_user_emails_dict
 from galaxy.schema.fields import LibraryFolderDatabaseIdField
 from galaxy.schema.schema import (
     BasicRoleModel,
@@ -114,9 +115,12 @@ class LibraryFoldersService(ServiceBase):
         #  Return roles that are available to select.
         elif scope == LibraryPermissionScope.available:
             roles, total_roles = trans.app.security_agent.get_valid_roles(trans, folder, query, page, page_limit)
+
+            private_role_emails = get_private_role_user_emails_dict(trans.sa_session)
             return_roles = []
             for role in roles:
-                return_roles.append(BasicRoleModel(id=role.id, name=role.name, type=role.type))
+                displayed_name = private_role_emails.get(role.id, role.name)
+                return_roles.append(BasicRoleModel(id=role.id, name=displayed_name, type=role.type))
             return LibraryAvailablePermissions(roles=return_roles, page=page, page_limit=page_limit, total=total_roles)
         else:
             raise RequestParameterInvalidException(

--- a/lib/galaxy/webapps/galaxy/services/roles.py
+++ b/lib/galaxy/webapps/galaxy/services/roles.py
@@ -21,8 +21,10 @@ def role_to_model(role, displayed_name: Optional[str] = None):
     item = role.to_dict(view="element")
     role_id = Security.security.encode_id(role.id)
     item["url"] = url_for("role", id=role_id)
+    # If displayed_name provided, use that value in place of Role.name. It is
+    # used to disambiguate generic role names like "private role".
     if displayed_name:
-        item["displayed_name"] = displayed_name
+        item["name"] = displayed_name
     return RoleModelResponse(**item)
 
 

--- a/lib/galaxy/webapps/galaxy/services/roles.py
+++ b/lib/galaxy/webapps/galaxy/services/roles.py
@@ -1,5 +1,8 @@
+from typing import Optional
+
 from galaxy.managers.context import ProvidesUserContext
 from galaxy.managers.roles import RoleManager
+from galaxy.model.db.role import get_private_role_user_emails_dict
 from galaxy.schema.fields import (
     DecodedDatabaseIdField,
     Security,
@@ -14,10 +17,12 @@ from galaxy.webapps.base.controller import url_for
 from galaxy.webapps.galaxy.services.base import ServiceBase
 
 
-def role_to_model(role):
+def role_to_model(role, displayed_name: Optional[str] = None):
     item = role.to_dict(view="element")
     role_id = Security.security.encode_id(role.id)
     item["url"] = url_for("role", id=role_id)
+    if displayed_name:
+        item["displayed_name"] = displayed_name
     return RoleModelResponse(**item)
 
 
@@ -33,7 +38,12 @@ class RolesService(ServiceBase):
 
     def get_index(self, trans: ProvidesUserContext) -> RoleListResponse:
         roles = self.role_manager.list_displayable_roles(trans)
-        return RoleListResponse(root=[role_to_model(r) for r in roles])
+        private_role_emails = get_private_role_user_emails_dict(trans.sa_session)
+        data = []
+        for role in roles:
+            displayed_name = private_role_emails.get(role.id, role.name)
+            data.append(role_to_model(role, displayed_name))
+        return RoleListResponse(root=data)
 
     def show(self, trans: ProvidesUserContext, id: DecodedDatabaseIdField) -> RoleModelResponse:
         role = self.role_manager.get(trans, id)

--- a/test/unit/data/model/conftest.py
+++ b/test/unit/data/model/conftest.py
@@ -439,7 +439,7 @@ def make_user_and_role(session, make_user, make_role, make_user_role_association
 
     def f(**kwd):
         user = make_user(**kwd)
-        private_role = make_role(type=m.Role.types.PRIVATE)
+        private_role = make_role(type=m.Role.types.PRIVATE, name=m.Role.default_name(m.Role.types.PRIVATE))
         make_user_role_association(user, private_role)
         return user, private_role
 

--- a/test/unit/data/model/db/test_role.py
+++ b/test/unit/data/model/db/test_role.py
@@ -1,5 +1,6 @@
 from galaxy.model import Role
 from galaxy.model.db.role import (
+    get_displayable_roles,
     get_npns_roles,
     get_private_user_role,
     get_roles_by_ids,
@@ -139,3 +140,22 @@ def test_get_valid_roles_exposed(session, make_user_and_role, make_user, make_ro
     search_query = "admin role%"
     roles = _get_valid_roles_exposed(session, search_query, is_admin, limit, page, page_limit)
     assert len(roles) == 0
+
+
+def test_get_displayable_roles(session, make_role, make_user_and_role):
+    # make users with private roles
+    user1, private_role1 = make_user_and_role(email="user1@example.com")
+    user2, private_role2 = make_user_and_role(email="user2@example.com")
+    # make 2 non-private roles, one is deleted, so it should not be returned
+    admin_role1 = make_role(type="admin", name="admin-role-1", description="Description of admin-role1")
+    make_role(type="admin", description="Description of admin-role1", deleted=True)
+
+    user_is_admin, security_agent = True, None
+    roles = get_displayable_roles(session, user1, user_is_admin, security_agent)
+    assert len(roles) == 3
+    assert roles[0].id == private_role1.id
+    assert roles[1].id == private_role2.id
+    assert roles[2].id == admin_role1.id
+    assert roles[0].name == "private role"
+    assert roles[1].name == "private role"
+    assert roles[2].name == "admin-role-1"

--- a/test/unit/data/model/db/test_role.py
+++ b/test/unit/data/model/db/test_role.py
@@ -2,6 +2,7 @@ from galaxy.model import Role
 from galaxy.model.db.role import (
     get_displayable_roles,
     get_npns_roles,
+    get_private_role_user_emails_dict,
     get_private_user_role,
     get_roles_by_ids,
 )
@@ -159,3 +160,19 @@ def test_get_displayable_roles(session, make_role, make_user_and_role):
     assert roles[0].name == "private role"
     assert roles[1].name == "private role"
     assert roles[2].name == "admin-role-1"
+
+
+def test_get_private_role_user_emails_dict(session, make_role, make_user_and_role):
+    # make users with private roles
+    user1, private_role1 = make_user_and_role(email="user1@example.com")
+    user2, private_role2 = make_user_and_role(email="user2@example.com")
+    user3, private_role3 = make_user_and_role(email="user3@example.com")
+    # make 2 non-private roles
+    make_role(type="admin", name="admin-role-1", description="Description of admin-role1")
+    make_role(type="admin", name="admin-role-2", description="Description of admin-role2")
+
+    data = get_private_role_user_emails_dict(session)
+    assert len(data) == 3  # only private role mappings are returned
+    assert data[private_role1.id] == "user1@example.com"
+    assert data[private_role2.id] == "user2@example.com"
+    assert data[private_role3.id] == "user3@example.com"

--- a/test/unit/data/model/db/test_security.py
+++ b/test/unit/data/model/db/test_security.py
@@ -13,7 +13,6 @@ from . import have_same_elements
 def test_private_user_role_assoc_not_affected_by_setting_user_roles(session, make_user_and_role):
     # Create user with a private role
     user, private_role = make_user_and_role()
-    assert user.email == private_role.name
     verify_user_associations(user, [], [private_role])  # the only existing association is with the private role
 
     # Delete user roles
@@ -25,7 +24,6 @@ def test_private_user_role_assoc_not_affected_by_setting_user_roles(session, mak
 def test_private_user_role_assoc_not_affected_by_setting_role_users(session, make_user_and_role):
     # Create user with a private role
     user, private_role = make_user_and_role()
-    assert user.email == private_role.name
     verify_user_associations(user, [], [private_role])  # the only existing association is with the private role
 
     # Update role users

--- a/test/unit/data/test_galaxy_mapping.py
+++ b/test/unit/data/test_galaxy_mapping.py
@@ -461,7 +461,7 @@ class TestMappings(BaseModelTestCase):
         def check_private_role(private_role, email):
             assert private_role.type == model.Role.types.PRIVATE
             assert len(private_role.users) == 1
-            assert private_role.name == email
+            assert private_role.name == model.Role.default_name(model.Role.types.PRIVATE)
 
         email = "rule_user_1@example.com"
         u = model.User(email=email, password="password")

